enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -229,24 +229,48 @@ The external network is used for external connections, such as `oc` clients.
 The internal network is used for internal connections, such as the ones from
 pods in the cluster.
 
-Choosing the internal network is a complex task, as it can collide with other
-IP ranges coming from DHCP or other environment configurations. In order to
-minimize risks, the API server service IP, which is always the [first one](https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/services.go#L47)
-from the service CIDR range, is selected by default.
+When configuring apiserver we need to provide an IP where it will be listening.
+This turns into an `EndpointSlice` object that will create an association
+between the apiserver service network IP and https port (defaults to
+10.43.0.1:443) and the IP where it will actually be listening on the real port,
+which defaults to 6443.
+Ovnk creates flows to turn this VIP into the one from the `EndpointSlice`. Here
+we need a valid IP, local to the node, where the apiserver is listening so that
+this translation succeeds.
+Imagine we use the same service network apiserver address for the
+`EndpointSlice`: 10.43.0.1. This creates a VIP-endpoint in ovnk that looks like
+10.43.0.1:443->10.43.0.1:6443. Imagine we have an interface in the host where
+the apiserver service network IP has been configured. While regular pod network
+origin traffic goes through the ovnk flows to translate it correctly, this is
+not the case with host network pods. These fall through the flow that redirects
+node IP traffic (which is the source for the host network pods) to the
+apiserver IP, 10.43.0.1, configured in some interface. Since the port is not
+modified by ovnk traffic ends up reaching whoever is listening on port 443 in
+any host interface. And there is a process listening there, the router for
+ingress traffic. Certificate validations fail, and even if they succeeded this
+is not the apiserver but a different service.
+
+In order to fix this we must use a different `EndpointSlice` that does not
+match the service network apiserver IP.
+
+From the service CIDR MicroShift can compute the next available subnet and use
+the first IP from the range as the new endpoint for apiserver. In case of
+defaults, service CIDR is 10.43.0.0/16 and the next subnet is 10.44.0.0. Since
+there is only 1 IP required the netmask can be at /32 and have only one
+available IP.
```

Comment:
Why not just use another IP on the service network directly?

Reply:
We need guaranteed 0% chance of collision.
One way is allocating a fake service having an empty/unavailable selector. This is a bit hacky: there would be a service somewhere in the cluster that does not map to any pod, its only there so no other service takes that IP.
Another way is using [ServiceIPStaticSubrange](https://kubernetes.io/blog/2022/05/23/service-ip-dynamic-and-static-allocation/#avoid-ClusterIP-conflict) feature gate. This requires enabling the feature gate and does not rule out collisions.
Also, the service CIDR is customizable, there could be deployments with small subnets and this takes up one IP.

Reply:
Yes, I think we need an IP before we get to the point where we can create a Service to allocate one, so that approach is out.

Can you capture these reasons in the body of the doc, maybe in the alternatives section?

Reply:
Done